### ISSUE TITLE
Extend ZAP API to support fuzzer 

### ISSUE THREAD

**Issue by eliasnester** [zaproxy/zaproxy#1689]
According to following document https://github.com/zaproxy/zaproxy/wiki/ApiGen_Index currently ZAP API doesn't support fuzzer. That would be great to extend API to allow doing fuzzing.

**Comment by thc202** [comment]
Please see related [OWASP ZAP User Group post](https://groups.google.com/d/msg/zaproxy-users/2f5IYbz1wZY/FhnKPqrj6wkJ) on why this was not added yet.
Looking forward feedback :)

**Comment by psiinon** [comment]
Yes, please let us know what you would like the fuzzing API to look like :)

**Comment by SyCode7** [comment]
Hi, just to check back if the fuzzer capacity was added.

**Comment by thc202** [comment]
Not yet, still waiting for feedback (per above comments).

**Comment by vincivis** [comment]
Hello everybody. Any news about this functionality?

**Comment by thc202** [comment]
No, still waiting for feedback.

**Comment by skzaproxy** [comment]
Hi, Is there any update on this requested functionality. ZAP is used widely and wish to have this functionality.

**Comment by kingthorin** [comment]
@skzaproxy no as per thc202's response way back in 2015, we're still waiting for users to give us some feedback/thoughts on how the fuzzer would be used and results leveraged via the API.

**Comment by Ali-Razmjoo** [comment]
Hello,
I was searching for fuzzer API and find this issue.
`Yes, please let us know what you would like the fuzzing API to look like :)`
The best way to know how to write this API is to find out the usual usages of ZAP fuzzer.

I personally use fuzzer to repeat requests, It's really awesome to use it for guessing usernames and passwords ( doing bruteforce with a list ) , I almost was able to do it for all of the CMS I ever need to guess password, Some of the CMS would redirect (return 302 code) after login, some of other still return code 200 but response is bigger or smaller size! ( I usually sort result with status code or response size)

In the other hand I sometimes use RCE or SQL Payloads on the web and checking the responses, some of the SQL Vulnerabilities (time base) always could be found by sorting response time without check the response content. 

So I think you just need to make this API just like the spider API (Show status, URLS, and all the options of fuzzer in a JSON) it would be best and with a simple script (I usually use python) user could be able analysis results in just some few seconds.

Example:
![brute](https://cloud.githubusercontent.com/assets/7676267/23518598/3bb5ad68-ff89-11e6-8c14-e73cfce8e41d.png)

there are some other problems or missing features in fuzzer which I found them by the experience of using ZAP these days. I'll share them in a new topic.

Regards.

**Comment by psiinon** [comment]
I now think the easiest way to work out what the API for any feature should look like is to imagine that the ZAP UI had to use it instead of having direct access to the classes.
In this case we have API calls for listing and accessing messages, so we'll need API calls to do things like:

- Specify the locations to attack
- Specify the payloads
- Page through the results
- Search for strings/regexes in the results
- etc etc

I think we should be able to define the fuzzing API on this basis, we just need someone with enough time to look into it:/

**Comment by Ali-Razmjoo** [comment]
Hey guys,

Since the issue created in 2015, I don't it's that complicated to add this feature. even it is, it could be done and by now many reported possible bugs were fixed. @psiinon said a few items that should be done so for the first step, let's finish the items, let me know if I miss something.

* define request structure (GET/POST/HEAD/ANY) (Could be select by Request ID in history and edit or read from)
```
GET http://127.0.0.1:5000/session/check?key=test&key2=test2 HTTP/1.1
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:59.0) Gecko/20100101 Firefox/59.0
Accept: text/plain, */*; q=0.01
Accept-Language: en-US,en;q=0.5
Referer: http://nettacker.z3r0d4y.com:5000/
X-Requested-With: XMLHttpRequest
Connection: keep-alive
Host: 127.0.0.1:5000
```
* mark the payloads location
```
GET http://127.0.0.1:5000/session/check?key={keylist_payload_1}&key2={keylist_payload_2} HTTP/1.1
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:59.0) Gecko/20100101 Firefox/59.0
Accept: text/plain, */*; q=0.01
Accept-Language: en-US,en;q=0.5
Referer: http://nettacker.z3r0d4y.com:5000/
X-Requested-With: XMLHttpRequest
Connection: keep-alive
Host: 127.0.0.1:5000
```
* define payloads Types (Empty/Null, File, File Fuzzers, Numberzz, Regex, Script, Strings)
* define payloads Content
* Multiline?
* Add Processor Type (Base64 Encode, Base64 Decode, Expand, JavaScript Escape, JavaScript Unescape, MD5 Hash, Postfix String, Prefix String, SHA-1 Hash, SHA-256 Hash, SHA-512 HASH, Script, Trim, URL Decode, URL Encode)
* Add Character Encoding (UTF8, ISO-8859-1, US-ASCII)
* Retries on IO error
* Limit maximm error?
* Max. errors allowed
* Payload replacement strategy (Depth First, Breadth First)
* Concurrent scanning threads per scan
* delay when fuzzing
* follow redirects?
* Message Processor (Fuzzer HTTP Processor, Payload Reflection Detector, Request Content-Length Updater, Tag Creator)
* **Start Fuzzer**
* Fuzzers Progress List (ID/Type/Name)
* Fuzzer Pause
* Fuzzer Stop
* Clean Finished Fuzzers
* Fuzzers Status
* Message sent
* Errors
* Show Errors
* Requests List (Task ID, Message Type, Req. Timestamp, Method, URL, Code, Reason, RTT, Size Req. Header, Size Req. Body, Size Resp. Header, Size Resp. Body, Highest Alert, State, Payloads)
* Show bytes only?
* Export

It's 27 items, let me know if I miss something?

I am not Java professional, and I didn't review OWASP ZAP code, but I don't think it gets more than week to add this for the core developers 😀 if someone guide me ( I am really not a document reader) and show me the entry points with a few examples, I can develop it in one thousand weeks.

any idea?

Regards.

**Comment by kingthorin** [comment]
@Ali-Razmjoo thanks for the response. This ticket is specifically for leveraging the Fuzzer from the Web API (I've updated the subject to make that more clear).

It's your second bullet and the end result that are the major hurdles. Having the user define 27 required/optional parameters times _n_ positions and expecting it to be  user friendly isn't really practical. Then there's the issue of how the results would be returned and then used? 
The results of fuzzing kind of require a human to review and decide what's important and what's an issue to investigate further. Yes a JSON (or whatever) blob of the results can be returned but in an automated context using the Web API how does that help or how would that be used?

**Comment by Ali-Razmjoo** [comment]
@kingthorin 

> It's your second bullet and the end result that are the major hurdles. Having the user define 27 required/optional parameters times n positions and expecting it to be user friendly isn't really practical. 
I was making a point to show it's not that complicated, 

>Yes a JSON (or whatever) blob of the results can be returned but in an automated context using the Web API how does that help or how would that be used?

it's called an API for a reason! if it was sposed to be human readable it was called something like UI. as a user I will decide how to manage the result (e.g. JSON Result) otherwise I will use the ZAP GUI.

>If you do fancy having a go then I'll be very happy to provide some pointers.
I do, I would like to develop any opensource projects, especially ZAP, I believe it's the greatest in its path!

Regards.

**Comment by Ali-Razmjoo** [comment]
I want to add a suggestion in here, In my new project (which is its start period) I setup a WebUI to demonstrate the API better! when the API starts, if anyone opens the index.html, they can use API in a nice UI (using bootstrap, jquery, etc..) or they can write their one app to interact with the API. It's really cool!

**Comment by psiinon** [comment]
Great :)
@Ali-Razmjoo - could you try to define the minimal set of API calls that would allow someone to do the absolute basics?
They should fit in with the rest of the ZAP API, but there isnt anything quite like this, so you'll have a fair amount of leeway.
We can then provide our feedback and once we have a good definition then we can help you translate that into code.
My initial thoughts are that there should be one API call to define a new fuzzing task. This would accept a full request to be fuzzed and return an index for the task. Future calls could then use this index and specify exactly what needs to be fuzzes and how.
@thc202 @kingthorin does that sound reasonable?

**Comment by kingthorin** [comment]
> I want to add a suggestion in here, In my new project (which is its start period) I setup a WebUI to demonstrate the API better! when the API starts, if anyone opens the index.html, they can use API in a nice UI (using bootstrap, jquery, etc..) or they can write their one app to interact with the API. It's really cool!

What's the suggestion? The Web API already has a web UI (granted it's very simple) just browse to it when ZAP is running. (http://localhost:8080 or whatever IP and port you have ZAP bound to).

@psiinon sounds like a good plan to me.

**Comment by Ali-Razmjoo** [comment]
@kingthorin I am aware of that, but it cannot be used by a normal human with that UI. my university website has better one :D

**Comment by Ali-Razmjoo** [comment]
@psiinon let me try, I will let you know if I've any question.

**Comment by thc202** [comment]
@psiinon I'd prefer having everything defined in one go than multiple endpoints to set up the fuzzer, keep it stateless.

**Comment by kingthorin** [comment]
@Ali-Razmjoo any progress?

**Comment by Ali-Razmjoo** [comment]
@kingthorin probably I didn't work on this yet, I was busy with planning for GSoC students, let you if I start working on this to prevent duplicate/conflicts. (Y)

**Comment by rakeshkmr516** [comment]
Hi @psiinon  
Thanks for implementing an awesome tool 
Could you please let me know any progress on fuzz API

**Comment by kingthorin** [comment]
There is no change. We'd be glad for your thoughts on a usable implementation though,

**Comment by davy320** [comment]
> Just a simple implementation I tried.

> Need some suggestions or changes.

![Screenshot from 2019-11-03 02-40-05](https://user-images.githubusercontent.com/14150099/68078660-637c3a00-fde3-11e9-892b-f902a6165d63.png)
And another idea, getting a JSON input like this: 
```

{
  "fuzzLocations": [
    {
      "location": "body",
      "start": 9,
      "end": 15,
      "payloads": [
        {
          "type": "file",
          "path": "/home/dennis/zaproxy-proj/zap-extensions/temp_payloads2.txt"
        }
      ]
    }
  ]
}

```

> These implementations for taking user inputs like this are in alpha phase. Any improvements or suggestions please comment.

**Comment by Ali-Razmjoo** [comment]
@davy320 looks cool, just need to add fuzz processor to payload section I guess.

**Comment by davy320** [comment]
@Ali-Razmjoo thanks for your feedback. The document below describes briefly what I was trying to do to make the user input convenient and keep the fuzzAPI stateless. The user guide vaguely explains the concept of JSON inputs. Please have a look and let me know how it is and if its ready for a pull request.

[OWASP ZAP FuzzAPI.pdf](https://github.com/zaproxy/zaproxy/files/3816187/OWASP.ZAP.FuzzAPI.pdf)

P.S. Was it okay for me to use the ZAP and OWASP logo in the document or is it against the license? Before publishing I will remove it.

**Comment by psiinon** [comment]
Hiya @davy320 - you can use the ZAP and OWASP logos in your document :)
Are you planning to submit a PR to this repo for your changes?

**Comment by davy320** [comment]
Hi @psiinon, I am planning to submit a pull request in a few days, after doing some more testing and checking for any issues that the user might encounter while using the API.
Waiting for more feedback :)

**Comment by rscottbailey** [comment]
I just ran across this topic today while researching the question of how to leverage ZAP for automated fuzz testing. Obviously API support would make that a lot easier. :)

With regard to the "but what about the output?" conversation above, which didn't seem to be definitely addressed anywhere, my initial thought (for whatever it might be worth) would be to have complementary "good"/"bad" regexs for response code, response headers, response body.

For example, I might consider responses 200 and 400 acceptable for a page, and any other response code would constitute "a failure".

I suppose it's also a "failure" if a response matches both "good" and "bad" regex, or neither (in the case where both are defined).

I hope this is useful! This is great stuff!

**Comment by rubiodamian** [comment]
any updates about this or an alternative workthrough?

**Comment by kingthorin** [comment]
You could grab the unfinished PR branch from 2019 and build it. Aside from that no there isn’t any updates as you can plainly see.

**Comment by binarymist** [comment]
I think what @rubiodamian was meaning was perhaps, Is there likely to be any progress on this in the near future from the contributors?

It's a good thing this is an open source project, so everyone with some programming skills is able and welcome to contribute as well as consume :-)

**Comment by rubiodamian** [comment]
@kingthorin i was just asking as @binarymist said.
This issue have 7 years open so i assumed that maybe an alternative solution,  using other parts of the API or something.

**Comment by kingthorin** [comment]
K, nope none of the core team are assigned, and no one has brought it up recently.

Yes it’s old but there are plenty of old issues and not enough time to get to everything.

As binarymist mentioned it is Open Source and we are happy for new contributions/contributors.

**Comment by psiinon** [comment]
Yep, would love someone to offer to work on it, but not something thats going to get to the top of my list anytime soon (unless something changes)...

**Comment by jja67** [comment]
hi there, i would like to use this Fuzz API even thought it is old and probably have a lot of old issues. but unfortuantely i can't even open the link https://github.com/zaproxy/zap-api-docs/fuzz , may i know where can i download the api and install into my ZAP?

**Comment by kingthorin** [comment]
There isn’t one. Read through the issue comments. The on-going debate is how it should work. 

As you’re reading you’ll notice there is a PR or two that attempt to implement some parts of it, you could pull one of those branches and custom build the fuzzer to use whatever API they’ve implemented, but that will obviously be unsupported by the core team.

**Comment by etkmac** [comment]
Hi, I found that there is no API yet for the fuzzer. I think one very useful could be to:
1. start by ID
2. pause by ID
3. pause all the running fuzzers
4. resume by ID
5. resume all the paused fuzzers
6. stop by ID
7. stop all the running fuzzers
8. and list the fuzzer tasks.

Just that I think it's simple and precise enough, and could help many of us.

**Comment by psiinon** [comment]
Thanks @etkmac :)
However its a bit more complicated than that :/
To start with there are no fuzzers, so you will not be able to start them by just providing an ID.
You will need to:

1. Specify a known request (by ID)
2. For each fuzz location:
- Define start an end offsets
- Define the fuzzer to use
- Specify any fuzzer parameters
